### ISSUE TITLE
loss function

### ISSUE THREAD

**Issue by WellYoungIOE** [ooooverflow/BiSeNet#5]
The loss function in the original paper is composed of two parts.  But you only use the output of feature fusion model to calculate the loss. And the loss they use is cross entropy, here you use is binary cross entropy.  Is there any reason to these changes? Thanks!

**Comment by WellYoungIOE** [comment]
`torch.nn.CrossEntropyLoss` is the one which combine  `nn.LogSoftmax()` and `nn.NLLLoss()`, not `torch.nn.BCEWithLogitsLoss()` you use in your code.
With auxiliary loss function, the output of the network should have three parts: output of feature fusion model, 1/8 and 1/16 output of context path. These three parts would be used to calculate three losses with ground truth separately. Then you should combine these three losses as final loss like the loss function did in the original paper. But in my experience, it would occupy a lot of GPU memory with pytorch during backward.
I'm not familiar with CamVid dataset, but if the label in CamVid does not have any overlap like person vs. woman. I strongly suggest you use cross entropy as the loss function.

**Comment by hubutui** [comment]
So in the article's Eq2, `lp` is the loss of the output of fusion feature, `li` is the loss of 1/8 and 1/16 output of context path. They also said that only auxiliary loss is used in training, so the params of spatial path is not trained?
Also, for 640x640 input image and resnet18 of context path, 1/8 and 1/16 output of context path is 1x256x40x40 and 1x512x20x20, upsample to 640x640 to calculate loss?

**Comment by WellYoungIOE** [comment]
Spatial path is only trained by `lp`.
1/8 and 1/16 output of context path is Bx128x80x80 and Bx256x40x40. I guess the author used 1x1 conv and upsample to make output with the same shape of label.

**Comment by hubutui** [comment]
Yeah, I found that https://github.com/CoinCheung/BiSeNet/blob/dev/train.py#L133 use upsample to get same shape of label, and compute loss. I still don't know that how they use auxiliary loss only for training phase.

**Comment by WellYoungIOE** [comment]
It seems that training with auxiliary loss like multi-task or multi-scale learning will occupy a lot of GPU memory. Because in pytorch, the loss are all calculated in the first GPU (gpu 0).  And that is why I came here to looking for the solution for training with auxiliary loss.

**Comment by hubutui** [comment]
Yeah, using dataparallel leads to unbalanced mem usage. Maybe we could try https://pytorch.org/docs/stable/distributed.html?

**Comment by hubutui** [comment]
@WellYoungIOE Did you use one hot encode provided in this repo? It would slow down speed.

**Comment by WellYoungIOE** [comment]
@hubutui I did not use that. I used my own code and the speed I mentioned is inference speed not training speed.

**Comment by hubutui** [comment]
@WellYoungIOE Emm, not sure why. But I inference at 20 imgs/s with 448x448 or 576x576. I think it's fast enough.

**Comment by hubutui** [comment]
I try resize my image to 1024x2048, and the inference speed is about 20 images/second with Titan Xp.

**Comment by WellYoungIOE** [comment]
@hubutui You mean 0.02s per frame?! Whoa! That fast? It seems that I need to change my GTX 1080Ti to Titan Xp.
But in the paper, they demonstrate their method can get 65.5 frames per second on 1024x2048 image. Your inference speed is still far away from that.

**Comment by hubutui** [comment]
Yeah, not sure how they got that performance. Maybe private deep learning framework?
